### ISSUE TITLE
CpaCyKeyGenHKDFOpData and CpaCyKeyHKDFCipherSuite in qat_hkdf.c are undefined

### ISSUE THREAD

**Issue by StHeck** [intel/QAT_Engine#146]
libtool: compile:  cc -DHAVE_CONFIG_H -I. -I.tools -I/usr/local/ssl/include -I/root/qatupdate/git/QAT/quickassist/utilities/libusdm_drv -I/root/qatupdate/git/QAT/quickassist/include -I/root/qatupdate/git/QAT/quickassist/include/lac -I/root/qatupdate/git/QAT/quickassist/lookaside/access_layer/include -fPIC -Wall -Wformat -Wformat-security -O2 -D_FORTIFY_SOURCE=2 -fstack-protector-strong -fno-strict-overflow -fno-delete-null-pointer-checks -fwrapv -DOPENSSL_ENABLE_QAT_UPSTREAM_DRIVER -DUSE_QAE_MEM -DOPENSSL_ENABLE_QAT_RSA -DOPENSSL_ENABLE_QAT_CIPHERS -DOPENSSL_ENABLE_QAT_ECDH -DOPENSSL_ENABLE_QAT_ECDSA -DOPENSSL_ENABLE_QAT_DSA -DOPENSSL_ENABLE_QAT_DH -DOPENSSL_ENABLE_QAT_PRF -DOPENSSL_ENABLE_QAT_HKDF -DOPENSSL_ENABLE_QAT_ECX -DOPENSSL_DISABLE_MULTIBUFF_RSA -DMULTIBUFF_HEURISTIC_TIMEOUT -g -O2 -MT qat_hkdf.lo -MD -MP -MF .deps/qat_hkdf.Tpo -c qat_hkdf.c  -fPIC -DPIC -o .libs/qat_hkdf.o
qat_hkdf.c:100:5: error: unknown type name 'CpaCyKeyGenHKDFOpData'
     CpaCyKeyGenHKDFOpData *hkdf_op_data;
     ^
qat_hkdf.c: In function 'qat_hkdf_init':
qat_hkdf.c:201:10: error: 'CpaCyKeyGenHKDFOpData' undeclared (first use in this function)
         (CpaCyKeyGenHKDFOpData *) qaeCryptoMemAlloc(sizeof(CpaCyKeyGenHKDFOpData), __FILE__,
          ^
qat_hkdf.c:201:10: note: each undeclared identifier is reported only once for each function it appears in
qat_hkdf.c:201:33: error: expected expression before ')' token
         (CpaCyKeyGenHKDFOpData *) qaeCryptoMemAlloc(sizeof(CpaCyKeyGenHKDFOpData), __FILE__,
                                 ^
qat_hkdf.c: In function 'qat_hkdf_cleanup':
qat_hkdf.c:250:39: error: request for member 'seedLen' in something not a structure or union
         if (qat_hkdf_ctx->hkdf_op_data->seedLen)
                                       ^
qat_hkdf.c:251:55: error: request for member 'seed' in something not a structure or union
             OPENSSL_cleanse(qat_hkdf_ctx->hkdf_op_data->seed,
                                                       ^
qat_hkdf.c:252:55: error: request for member 'seedLen' in something not a structure or union
                             qat_hkdf_ctx->hkdf_op_data->seedLen);
                                                       ^
qat_hkdf.c:254:39: error: request for member 'secretLen' in something not a structure or union
         if (qat_hkdf_ctx->hkdf_op_data->secretLen)
                                       ^
qat_hkdf.c:255:55: error: request for member 'secret' in something not a structure or union
             OPENSSL_cleanse(qat_hkdf_ctx->hkdf_op_data->secret,
                                                       ^
qat_hkdf.c:256:55: error: request for member 'secretLen' in something not a structure or union
                             qat_hkdf_ctx->hkdf_op_data->secretLen);
                                                       ^
qat_hkdf.c:258:39: error: request for member 'infoLen' in something not a structure or union
         if (qat_hkdf_ctx->hkdf_op_data->infoLen)
                                       ^
qat_hkdf.c:259:55: error: request for member 'info' in something not a structure or union
             OPENSSL_cleanse(qat_hkdf_ctx->hkdf_op_data->info,
                                                       ^
qat_hkdf.c:260:55: error: request for member 'infoLen' in something not a structure or union
                             qat_hkdf_ctx->hkdf_op_data->infoLen);
                                                       ^
qat_hkdf.c: In function 'qat_hkdf_ctrl':
qat_hkdf.c:342:55: error: request for member 'seed' in something not a structure or union
             OPENSSL_cleanse(qat_hkdf_ctx->hkdf_op_data->seed,
                                                       ^
qat_hkdf.c:343:55: error: request for member 'seedLen' in something not a structure or union
                             qat_hkdf_ctx->hkdf_op_data->seedLen);
                                                       ^
qat_hkdf.c:344:39: error: request for member 'seedLen' in something not a structure or union
             qat_hkdf_ctx->hkdf_op_data->seedLen = 0;
                                       ^
qat_hkdf.c:346:46: error: request for member 'seed' in something not a structure or union
             memcpy(qat_hkdf_ctx->hkdf_op_data->seed, p2, p1);
                                              ^
qat_hkdf.c:347:39: error: request for member 'seedLen' in something not a structure or union
             qat_hkdf_ctx->hkdf_op_data->seedLen = p1;
                                       ^
qat_hkdf.c:361:55: error: request for member 'secret' in something not a structure or union
             OPENSSL_cleanse(qat_hkdf_ctx->hkdf_op_data->secret,
                                                       ^
qat_hkdf.c:362:55: error: request for member 'secretLen' in something not a structure or union
                             qat_hkdf_ctx->hkdf_op_data->secretLen);
                                                       ^
qat_hkdf.c:363:39: error: request for member 'secretLen' in something not a structure or union
             qat_hkdf_ctx->hkdf_op_data->secretLen = 0;
                                       ^
qat_hkdf.c:365:46: error: request for member 'secret' in something not a structure or union
             memcpy(qat_hkdf_ctx->hkdf_op_data->secret, p2, p1);
                                              ^
qat_hkdf.c:366:39: error: request for member 'secretLen' in something not a structure or union
             qat_hkdf_ctx->hkdf_op_data->secretLen = p1;
                                       ^
qat_hkdf.c:378:87: error: request for member 'infoLen' in something not a structure or union
             if (p1 < 0 || p1 > (int) QAT_HKDF_INFO_MAXBUF - qat_hkdf_ctx->hkdf_op_data->infoLen) {
                                                                                       ^
qat_hkdf.c:383:46: error: request for member 'info' in something not a structure or union
             memcpy(qat_hkdf_ctx->hkdf_op_data->info
                                              ^
qat_hkdf.c:384:48: error: request for member 'infoLen' in something not a structure or union
                    + qat_hkdf_ctx->hkdf_op_data->infoLen, p2, p1);
                                                ^
qat_hkdf.c:385:39: error: request for member 'infoLen' in something not a structure or union
             qat_hkdf_ctx->hkdf_op_data->infoLen += p1;
                                       ^
qat_hkdf.c: At top level:
qat_hkdf.c:434:33: error: unknown type name 'CpaCyKeyHKDFCipherSuite'
                                 CpaCyKeyHKDFCipherSuite* cipher_suite)
                                 ^
qat_hkdf.c: In function 'qat_set_hkdf_mode':
qat_hkdf.c:483:39: error: request for member 'hkdfKeyOp' in something not a structure or union
             qat_hkdf_ctx->hkdf_op_data->hkdfKeyOp = CPA_CY_HKDF_KEY_EXTRACT;
                                       ^
qat_hkdf.c:483:53: error: 'CPA_CY_HKDF_KEY_EXTRACT' undeclared (first use in this function)
             qat_hkdf_ctx->hkdf_op_data->hkdfKeyOp = CPA_CY_HKDF_KEY_EXTRACT;
                                                     ^
qat_hkdf.c:487:39: error: request for member 'hkdfKeyOp' in something not a structure or union
             qat_hkdf_ctx->hkdf_op_data->hkdfKeyOp = CPA_CY_HKDF_KEY_EXPAND;
                                       ^
qat_hkdf.c:487:53: error: 'CPA_CY_HKDF_KEY_EXPAND' undeclared (first use in this function)
             qat_hkdf_ctx->hkdf_op_data->hkdfKeyOp = CPA_CY_HKDF_KEY_EXPAND;
                                                     ^
qat_hkdf.c:491:39: error: request for member 'hkdfKeyOp' in something not a structure or union
             qat_hkdf_ctx->hkdf_op_data->hkdfKeyOp = CPA_CY_HKDF_KEY_EXTRACT_EXPAND;
                                       ^
qat_hkdf.c:491:53: error: 'CPA_CY_HKDF_KEY_EXTRACT_EXPAND' undeclared (first use in this function)
             qat_hkdf_ctx->hkdf_op_data->hkdfKeyOp = CPA_CY_HKDF_KEY_EXTRACT_EXPAND;
                                                     ^
qat_hkdf.c: In function 'qat_hkdf_derive':
qat_hkdf.c:521:5: error: unknown type name 'CpaCyKeyHKDFCipherSuite'
     CpaCyKeyHKDFCipherSuite cipher_suite;
     ^
qat_hkdf.c:553:10: warning: implicit declaration of function 'qat_get_cipher_suite' [-Wimplicit-function-declaration]
     if (!qat_get_cipher_suite(qat_hkdf_ctx, &cipher_suite)) {
          ^
qat_hkdf.c:578:35: error: request for member 'hkdfKeyOp' in something not a structure or union
     if (qat_hkdf_ctx->hkdf_op_data->hkdfKeyOp == CPA_CY_HKDF_KEY_EXTRACT_EXPAND) {
                                   ^
qat_hkdf.c:578:50: error: 'CPA_CY_HKDF_KEY_EXTRACT_EXPAND' undeclared (first use in this function)
     if (qat_hkdf_ctx->hkdf_op_data->hkdfKeyOp == CPA_CY_HKDF_KEY_EXTRACT_EXPAND) {
                                                  ^
qat_hkdf.c:649:18: warning: implicit declaration of function 'cpaCyKeyGenTls3' [-Wimplicit-function-declaration]
         status = cpaCyKeyGenTls3(qat_instance_handles[inst_num],
                  ^
Makefile:729: recipe for target 'qat_hkdf.lo' failed
make[1]: *** [qat_hkdf.lo] Error 1

**Comment by paulturx** [comment]
Hi StHeck
These structures are declared in QAT driver header file quickassist/include/lac/cpa_cy_key.h, a header file that is included in source file qat_hkdf.c. The '-I/root/qatupdate/git/QAT/quickassist/include/lac' you have pasted in your comment above should have provided the path to this header file.  Could you paste the configure command you used to build the QAT engine please?

Regards,
    paulturx

**Comment by zhangcunli** [comment]
Same error:
qat_hkdf.c:100:5: error: unknown type name 'CpaCyKeyGenHKDFOpData'
     CpaCyKeyGenHKDFOpData *hkdf_op_data;
>>>>>>>>>>>>>>>>>>>>

I download the Intel_QAT_Linux_1.7.l.4.5.0-00034.zip from "https://downloadcenter.intel.com/download/28797?v=t"；
But I can't find the struct CpaCyKeyGenHKDFOpData in quickassist/include/lac/cpa_cy_key.h.

**Comment by StHeck** [comment]
For Ubuntu Xenial (current work), we use qat1.7.l.4.3.0-00033. cpa_cy_key.h is not in that release.

For Ubuntu Bionic (future work), we use qat1.7.l.4.7.0-00006. cpa_cy_key.h is in that release.

We had just been cloning from the master for QAT_Engine. We are now cloning the previous tag, v0.5.43 for our current work.

So, it looks like it's just a QAT_Engine/driver version sync issue.

All good now.

Thanks.

________________________________
